runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/Networks.java

```diff
@@ -156,21 +169,34 @@ public final NodeT apply(NodeT input) {
 
     // Linked hashset will prevent duplicates from appearing and will maintain insertion order.
     LinkedHashSet<NodeT> nodes = new LinkedHashSet<>(network.nodes().size());
+    // Sort all of the nodes before adding them to the processing order, so we get a deterministic
+    // order. Take an explicit comparator.
+    List<NodeT> roots = new ArrayList<>();
```

Comment:
Consider using the NetworkBuilder.from(network).withNodeOrder(nodeSubOrder) instead of sorting the successor sets yourself.

Reply:
Done